transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,142 @@
+
+    private final AtomicInteger acquiredChannelCount = new AtomicInteger();
+    private final Queue<AcquireTask<C, K>> pendingAcquireQueue = new ConcurrentLinkedQueue<AcquireTask<C, K>>();
+    private final FutureListener<C> decrementListener = new FutureListener<C>() {
+        @Override
+        public void operationComplete(Future<C> future) throws Exception {
+            if (future.isSuccess()) {
+                future.getNow().closeFuture().addListener(closeListener);
+            } else {
+                acquiredChannelCount.decrementAndGet();
+                runTaskQueue();
+            }
+        }
+    };
+    private final FutureListener<Boolean> runListener = new FutureListener<Boolean>() {
+        @Override
+        public void operationComplete(Future<Boolean> future) throws Exception {
+            runTaskQueue();
+        }
+    };
+    private final ChannelFutureListener closeListener = new ChannelFutureListener() {
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            // The channel was closed so try to process the taskqueue again as this means we lost one connection out
+            // of the pool.
+            runTaskQueue();
+        }
+    };
+
+    private final ChannelPool<C, K> pool;
+    private final int maxConnections;
+
+    public FixedChannelPool(ChannelPool<C, K> pool, int maxConnections) {
+        if (maxConnections < 1) {
+            throw new IllegalArgumentException("maxConnections must be >= 1 but was " + maxConnections);
+        }
+        this.pool = checkNotNull(pool, "pool");
+        this.maxConnections = maxConnections;
+    }
+
+    @Override
+    public Future<C> acquire(K key) {
+        return acquire(key, ImmediateEventExecutor.INSTANCE.<C>newPromise());
+    }
+
+    @Override
+    public Future<Boolean> release(C channel) {
+        return release(channel, ImmediateEventExecutor.INSTANCE.<Boolean>newPromise());
+    }
+
+    @Override
+    public Future<C> acquire(K key, Promise<C> promise) {
+        if (acquiredChannelCount.incrementAndGet() <= maxConnections) {
+            promise.addListener(decrementListener);
+            return pool.acquire(key, promise);
+        }
+        pendingAcquireQueue.add(new AcquireTask<C, K>(key, promise));
+        return promise;
+    }
+
+    @Override
+    public Future<Boolean> release(C channel, Promise<Boolean> promise) {
+        if (channel.isActive()) {
+            Future<Boolean> f = pool.release(channel, promise);
+            if (f.isDone()) {
+                runTaskQueue();
+            } else {
+                f.addListener(runListener);
+            }
+            return f;
+        } else {
+            return promise.setSuccess(Boolean.FALSE);
+        }
+    }
+
+    private void runTaskQueue() {
+        for (;;) {
+            AcquireTask<C, K> task = pendingAcquireQueue.peek();
```

Comment:
I believe ``runTaskQueue`` can be called from any thread since it's called from a channel listener.  I see you're using ``ImmediateEventExecutor`` which I'm not familiar with so I could be wrong but if I'm reading it right it means the thread performing the operation on the channel will end up calling the listener which then calls this method.

Since the channels are unrelated I think that:

- You can execute the same task more than once from multiple threads because ``peek`` is used
- You can get exceptions on line 122 due to the queue being emptied by another thread